### ISSUE TITLE
[pycde] ESI error

### ISSUE THREAD

**Issue by Dragon-Git** [llvm/circt#5490]
When I use pycde.esi.DeclareRandomAccessMemory to generate multi-port ram, it produces the following error:

```
Assertion failed: toServerReq && toClientReq, file C:\__w\1\s\CIRCT\circt\lib\Dialect\ESI\ESIServices.cpp, line 171
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Exception Code: 0x80000003
0x00007FFA4B0865F5, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\_circt.cp311-win_amd64.pyd(0x00007FFA4B040000) + 0x465F5 byte(s), PyInit__circt() + 0x3A3C5 byte(s)
0x00007FFA68F01881, C:\WINDOWS\System32\ucrtbase.dll(0x00007FFA68E90000) + 0x71881 byte(s), raise() + 0x1E1 byte(s)       
0x00007FFA68F02851, C:\WINDOWS\System32\ucrtbase.dll(0x00007FFA68E90000) + 0x72851 byte(s), abort() + 0x31 byte(s)        
0x00007FFA68F0426E, C:\WINDOWS\System32\ucrtbase.dll(0x00007FFA68E90000) + 0x7426E byte(s), _get_wpgmptr() + 0x195E byte(s)
0x00007FFA68F04165, C:\WINDOWS\System32\ucrtbase.dll(0x00007FFA68E90000) + 0x74165 byte(s), _get_wpgmptr() + 0x1855 byte(s)
0x00007FFA68F044F1, C:\WINDOWS\System32\ucrtbase.dll(0x00007FFA68E90000) + 0x744F1 byte(s), _wassert() + 0x71 byte(s)     
0x00007FF9EF579EBB, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x209EBB byte(s), mlirRegisterTransformsViewOpGraph() + 0x1E853B byte(s)
0x00007FF9EF56FF0E, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x1FFF0E byte(s), mlirRegisterTransformsViewOpGraph() + 0x1DE58E byte(s)
0x00007FF9EF57C8C9, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x20C8C9 byte(s), mlirRegisterTransformsViewOpGraph() + 0x1EAF49 byte(s)
0x00007FF9EF57B926, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x20B926 byte(s), mlirRegisterTransformsViewOpGraph() + 0x1E9FA6 byte(s)
0x00007FF9EF57CB63, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x20CB63 byte(s), mlirRegisterTransformsViewOpGraph() + 0x1EB1E3 byte(s)
0x00007FF9EF4BE527, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x14E527 byte(s), mlirRegisterTransformsViewOpGraph() + 0x12CBA7 byte(s)
0x00007FF9EF4C3FB7, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x153FB7 byte(s), mlirRegisterTransformsViewOpGraph() + 0x132637 byte(s)
0x00007FF9EF4C5A64, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x155A64 byte(s), mlirRegisterTransformsViewOpGraph() + 0x1340E4 byte(s)
0x00007FF9EF4C4730, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x154730 byte(s), mlirRegisterTransformsViewOpGraph() + 0x132DB0 byte(s)
0x00007FF9EF385DE1, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\PyCDE_CIRCTPythonCAPI.dll(0x00007FF9EF370000) + 0x15DE1 byte(s), mlirPassManagerRunOnOp() + 0x11 byte(s)
0x00007FFA3591ABE8, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\_mlir.cp311-win_amd64.pyd(0x00007FFA35820000) + 0xFABE8 byte(s), PyInit__mlir() + 0xEC268 byte(s)
0x00007FFA3591AD04, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\_mlir.cp311-win_amd64.pyd(0x00007FFA35820000) + 0xFAD04 byte(s), PyInit__mlir() + 0xEC384 byte(s)
0x00007FFA3582C53A, C:\Users\uname\project\env\pycde_example\Lib\site-packages\pycde\circt\_mlir_libs\_mlir.cp311-win_amd64.pyd(0x00007FFA35820000) + 0xC53A byte(s)
0x00007FFA03DE051C, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x1051C byte(s), _PyObject_MakeTpCall() + 0x26C byte(s)
0x00007FFA03DE4548, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x14548 byte(s), PyObject_Vectorcall() + 0x1E8 byte(s)
0x00007FFA03DE5DA4, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x15DA4 byte(s), _PyEval_EvalFrameDefault() + 0x784 byte(s)
0x00007FFA03E02F73, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x32F73 byte(s), PyMapping_Check() + 0x1EB byte(s)
0x00007FFA03E027DB, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x327DB byte(s), PyEval_EvalCode() + 0x97 byte(s)
0x00007FFA03E72582, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0xA2582 byte(s), _Py_SourceAsString() + 0x646 byte(s)
0x00007FFA03E724FE, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0xA24FE byte(s), _Py_SourceAsString() + 0x5C2 byte(s)
0x00007FFA03FB3DF4, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x1E3DF4 byte(s), PyThread_tss_is_created() + 0x4EA14 byte(s)
0x00007FFA03F16B99, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x146B99 byte(s), _PyRun_SimpleFileObject() + 0x11D byte(s)
0x00007FFA03F178E0, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x1478E0 byte(s), _PyRun_AnyFileObject() + 0x54 byte(s)
0x00007FFA03F17387, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x147387 byte(s), PyDict_DelItemString() + 0x6FF byte(s)
0x00007FFA03F17243, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x147243 byte(s), PyDict_DelItemString() + 0x5BB byte(s)
0x00007FFA03ED0D80, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x100D80 byte(s), Py_RunMain() + 0x368 byte(s)
0x00007FFA03ED0A2D, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python311.dll(0x00007FFA03DD0000) + 0x100A2D byte(s), Py_RunMain() + 0x15 byte(s)
0x00007FF65B8E4320, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python3.11.exe(0x00007FF65B8E0000) + 0x4320 byte(s)
0x00007FF65B8E58D4, C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1264.0_x64__qbz5n2kfra8p0\python3.11.exe(0x00007FF65B8E0000) + 0x58D4 byte(s)
0x00007FFA69A67614, C:\WINDOWS\System32\KERNEL32.DLL(0x00007FFA69A50000) + 0x17614 byte(s), BaseThreadInitThunk() + 0x14 byte(s)
0x00007FFA6B0A26F1, C:\WINDOWS\SYSTEM32\ntdll.dll(0x00007FFA6B050000) + 0x526F1 byte(s), RtlUserThreadStart() + 0x21 byte(s)
```

OS: Window 10
pycde version: 0.2.1.dev823
The python code is:

```python
from pycde import (Clock, Input, Module, System, generator, types)
from pycde.constructs import Wire
from pycde import esi

RamI32x32 = esi.DeclareRandomAccessMemory(types.i32, 32, "RamI32x32")
WriteType = RamI32x32.write.to_server_type

@esi.ServiceDecl
class MP_RAM_io:
    write = esi.FromServer(WriteType)
    read = esi.ToFromServer(to_server_type=types.i32, to_client_type=types.i5)
    read1 = esi.ToFromServer(to_server_type=types.i32, to_client_type=types.i5)

class MP_RAM(Module):
    clk = Clock()
    rst = Input(types.i1)

    @generator
    def construct(io):
        RamI32x32.write(MP_RAM_io.write("write"))
        read_address = Wire(RamI32x32.read.to_server_type)
        read_data = RamI32x32.read(read_address)
        read_address.assign(MP_RAM_io.read(read_data, "read"))

        read1_address = Wire(RamI32x32.read.to_server_type)
        read1_data = RamI32x32.read(read1_address)
        read1_address.assign(MP_RAM_io.read1(read1_data, "read1"))

        RamI32x32.instantiate_builtin("sv_mem", [], inputs=[io.clk, io.rst])

mod = System([MP_RAM],name="test_mp_ram")
mod.compile()

```

**Comment by teqdruid** [comment]
Confirmed bug. It's a result of the poor implementation of `ToFromServer` requests paired with my own laziness. The quick fix is to add a name to one or both of the `read`s:

```
        read_address = Wire(RamI32x32.read.to_server_type)
        read_data = RamI32x32.read(read_address, "read0")
        read_address.assign(MP_RAM_io.read(read_data, "read"))
```

TL;DR: there's no such thing as a bi-directional channel in ESI, so `ToFromServer` requests get broken up early on then re-joined based on name later. If there's anything which prevents determining those pairs, it bails.

**Comment by Dragon-Git** [comment]
Thank you for providing the solution! I have tried your suggestion and successfully resolved my issue. I really appreciate your help!
